### PR TITLE
breaking, fix: guild_member state tracking

### PR DESCRIPTION
**Pull request by braindigitalis**
- [x] breaking, fix: guild_member state tracking.
    You must now use add_role, remove_role or set_roles, also set_nickname on a guild_member for edit_guild_mebmer to have any effect on that value. This fixes longstanding issue where more permissions are required than should be, to edit guild members
- [x] test: move mutltiheader cookie test to extended, impossible to test in uk and eu due to GDPR

## Code change checklist

- [x] I have ensured that all methods and functions are **fully documented** using doxygen style comments.
- [x] My code follows the [coding style guide](https://dpp.dev/coding-standards.html).
- [x] I tested that my change works before raising the PR.
- [x] I have ensured that I did not break any existing API calls.
- [x] I have not built my pull request using AI, a static analysis tool or similar without any human oversight.

### CODE REVIEW

File: src/dpp/guild.cpp [hunk 1]
```diff
@@ -112,6 +112,27 @@ std::string guild_member::get_mention() const {
...
+}
+
+guild_member& guild_member::add_role(dpp::snowflake role_id) {
+	roles.emplace_back(role_id);
+	flags |= gm_roles_action;
+	return *this;
+}
+
+guild_member& guild_member::remove_role(dpp::snowflake role_id) {
+	roles.erase(std::remove_if(roles.begin(), roles.end(), [&role_id](dpp::snowflake role) {
```
Comment (Mishura4) [inline src/dpp/guild.cpp:126]: Can just use `std::remove` here instead of using `std::remove_if` with a lambda

File: src/dpp/guild.cpp [hunk 2]
```diff
@@ -211,19 +232,23 @@ std::string guild_member::build_json(bool with_id) const {
...
 	}
 
-	if (!this->roles.empty()) {
-		j["roles"] = {};
-		for (auto & role : this->roles) {
-			j["roles"].push_back(std::to_string(role));
+	if (this->flags & gm_roles_action) {
+		if (!this->roles.empty()) {
+			j["roles"] = {};
+			for (auto & role : this->roles) {
```
Comment (Mishura4) [inline src/dpp/guild.cpp:246]: `const auto&`!

File: src/dpp/guild.cpp [hunk 3]
```diff
@@ -211,19 +232,23 @@ std::string guild_member::build_json(bool with_id) const {
...
+			j["nick"] = json::value_t::null;
+		}
 	}
 
-	if (!this->roles.empty()) {
-		j["roles"] = {};
-		for (auto & role : this->roles) {
-			j["roles"].push_back(std::to_string(role));
+	if (this->flags & gm_roles_action) {
+		if (!this->roles.empty()) {
```
Comment (Mishura4) [inline src/dpp/guild.cpp:244]: This whole branch is redundant, the `for` loop will be a no-op anyways if it's empty

File: include/dpp/guild.h [hunk 4]
```diff
@@ -378,6 +380,40 @@ class DPP_EXPORT guild_member {
...
+	 * 
+	 * @param role_id Role ID to remove
+	 * @return guild_member& Reference to self
+	 */
+	guild_member& remove_role(dpp::snowflake role_id);
+
+	/**
+	 * @brief Set a new role list for this member
+	 * @note This call sets the role change bit, which causes the new role
+	 * list to be sent if this is passed to dpp::clusterguild_edit_member
```
Comment (Mishura4) [inline include/dpp/guild.h:409]: Typo in the documentation: `dpp::clusterguild_edit_member`

File: include/dpp/guild.h [hunk 5]
```diff
@@ -378,6 +380,40 @@ class DPP_EXPORT guild_member {
...
+	 * 
+	 * @param role_id Role ID to add
+	 * @return guild_member& Reference to self
+	 */
+	guild_member& add_role(dpp::snowflake role_id);
+
+	/**
+	 * @brief Remove a role from this member
+	 * @note This call sets the role change bit, which causes the new role
+	 * list to be sent if this is passed to dpp::clusterguild_edit_member
```
Comment (Mishura4) [inline include/dpp/guild.h:398]: Typo in the documentation: `dpp::clusterguild_edit_member`

**Review by Mishura4 — CHANGES_REQUESTED** [review]
Request changes + making the fields private as we discussed on Discord, and good for me

File: include/dpp/guild.h [hunk 6]
```diff
@@ -180,6 +180,8 @@ enum guild_member_flags : uint16_t {
 	gm_bypasses_verification = 0b0000000010000000,
 	/** Member has started onboarding */
 	gm_started_onboarding = 0b0000000100000000,
+	gm_roles_action    = 0b0000001000000000,
```
Comment (Mishura4) [inline include/dpp/guild.h:183]: Will this be sent to Discord? Will that cause issues?

Comment (braindigitalis) [reply include/dpp/guild.h:183]: no, we serialise these into guild member data ourselves, our bitfields don't match discord's definitions

File: src/dpp/guild.cpp (hunk 1 above)
Comment (braindigitalis) [reply src/dpp/guild.cpp:126]: did you mean std::erase, seems to be C++20 only?

https://en.cppreference.com/w/cpp/container/vector/erase2

File: src/dpp/guild.cpp (hunk 3 above)
Comment (Mishura4) [reply src/dpp/guild.cpp:253]: Huh
I am talking about this->roles yes
You check if it's empty then you loop over it to push_back roles in it, the loop is also a check that it's empty

File: src/dpp/role.cpp [hunk 7]
```diff
@@ -413,8 +413,9 @@ members_container role::get_members() const {
 		}
 		for (auto & m : g->members) {
 			/* Iterate all members and use std::find on their role list to see who has this role */
-			auto i = std::find(m.second.roles.begin(), m.second.roles.end(), this->id);
-			if (i != m.second.roles.end()) {
+			auto r = m.second.get_roles();
```
Comment (Mishura4) [inline src/dpp/role.cpp:416]: Possibly an unintended copy here : you probably want `const auto&`

File: src/dpp/guild.cpp [hunk 8]
```diff
@@ -211,19 +239,19 @@ std::string guild_member::build_json(bool with_id) const {
...
+		} else {
+			j["nick"] = json::value_t::null;
+		}
 	}
 
-	if (!this->roles.empty()) {
-		j["roles"] = {};
-		for (auto & role : this->roles) {
+	if (gm_roles_action) {
+    		j["roles"] = {};
```
Comment (Mishura4) [inline src/dpp/guild.cpp:251]: Space indent :omg:

File: src/dpp/guild.cpp [hunk 9]
```diff
@@ -123,9 +123,7 @@ guild_member& guild_member::add_role(dpp::snowflake role_id) {
 }
 
 guild_member& guild_member::remove_role(dpp::snowflake role_id) {
-	roles.erase(std::remove_if(roles.begin(), roles.end(), [&role_id](dpp::snowflake role) {
-		return role == role_id;
-	}), roles.end());
+	std::remove(roles.begin(), roles.end(), role_id);
```
Comment (Mishura4) [inline src/dpp/guild.cpp:126]: Should be `roles.erase(std::remove(roles.begin(), roles.end(), role_id));`

File: src/dpp/guild.cpp [hunk 10]
```diff
@@ -112,6 +112,34 @@ std::string guild_member::get_mention() const {
...
+	return *this;
+}
+
+guild_member& guild_member::remove_role(dpp::snowflake role_id) {
+	roles.erase(std::remove(roles.begin(), roles.end(), role_id), roles.end());
+	flags |= gm_roles_action;
+	return *this;
+}
+
+std::string guild_member::get_nickname() {
```
Comment (Mishura4) [inline src/dpp/guild.cpp:131]: Shouldn't this return a const reference? also, shouldn't the method be const-qualified?

File: src/dpp/guild.cpp [hunk 11]
```diff
@@ -112,6 +112,34 @@ std::string guild_member::get_mention() const {
...
+	return nickname;
+}
+
+std::vector<dpp::snowflake> guild_member::get_roles() const {
```
Comment (Mishura4) [inline src/dpp/guild.cpp:135]: Shouldn't this return a const reference?

**Review by Mishura4 — CHANGES_REQUESTED** [review]

**Review by Jaskowicz1 — APPROVED** [review]
Bellissimo,

**Review by Mishura4 — APPROVED** [review]
Toad doing a thumbs up motion

**Pull request merged**